cmd/process-user-data/provision.go

```diff
@@ -45,21 +45,53 @@ func isAzure(ctx context.Context) bool {
 	return resp.StatusCode == http.StatusOK
 }
 
-// Get the URL to retrieve the userData from the instance metadata service
-func getUserDataURL(ctx context.Context) string {
-	userDataUrl := ""
+// Add a method to check if the VM is running on AWS
+// by checking if the AWS IMDS endpoint is reachable
+// If the VM is running on AWS, return true
+func isAWS(ctx context.Context) bool {
+
+	// Create a new HTTP client
+	client := &http.Client{}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, AWSImdsUrl, nil)
+	if err != nil {
+		fmt.Printf("failed to create request: %s\n", err)
+		return false
+	}
+
+	// Send the request and retrieve the response
+	resp, err := client.Do(req)
+	if err != nil {
+		fmt.Printf("failed to send request: %s\n", err)
+		return false
+	}
+	defer resp.Body.Close()
+
+	// Check if the response was successful
+	return resp.StatusCode == http.StatusOK
+}
+
+// Get the provider and the URL to retrieve the userData from the instance metadata service
+func getProviderAndUserDataURL(ctx context.Context) (provider string, userDataUrl string) {
 
 	if isAzure(ctx) {
+		provider = providerAzure
 		// If the VM is running on Azure, retrieve the userData from the Azure IMDS endpoint
 		userDataUrl = AzureUserDataImdsUrl
 	}
 
-	return userDataUrl
+	if isAWS(ctx) {
```

Comment:
We'll need heuristics if using generic qcow2/raw images. But if we are building native images, then this heuristic is not needed and probably set via ldflags. Should I explore this approach as well in this PR or a next one ?

Reply:
I see. We maybe want to employ some other heuristics to avoid the http roundtrips:

there are a set of [methods for AWS](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/identify_ec2_instances.html)

for azure the prescribes dmidecode method via smbios should also work: `sudo /usr/sbin/dmidecode -s {system-product-name,system-manufactorer}`. We could possibly retrieve this by using a go library: https://blog.gopheracademy.com/advent-2017/accessing-smbios-information-with-go/

Reply:
The dmidecode parser code has not been updated since many years. Other option is to directly parse `/sys` to get some of these info. But reading the document, this is also not fool proof.
As for AWS, the definitive way is to get the `instance identity document` from the reference doc you mentioned. I'm using that to check if it's AWS. Likewise for Azure, using what is described in https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux

@mkulke I'm thinking of continuing with additional exploration to use `/sys` or compile time option for provider specific builds in a future PR. Hope this is fine with you

Reply:
Yes, let's open an issue as a reminder about the concern with the current approach, though.

Reply:
Yes, let's open an issue as a reminder about the concern with the current approach, though.